rfcs/plugin-api/RFC.md

```diff
@@ -0,0 +1,279 @@
+            sdk: {
+              languageNegotiation: {
+                strategies: [{ type: "localStorage" }, { type: "navigator" }],
+              },
+            }
+          }
+        }),
+      )
+    )
+}
+```
+
+❌ = dealbreaker
+
+|         Pros         |                       Cons                        |
+| :------------------: | :-----------------------------------------------: |
+|        Simple        | Every plugin takes entire core config as argument |
+| Backwards compatible |            Nesting becomes convoluted             |
+|                      |               no error handling ❌                |
+|                      |        unclear when to execute a plugin ❌        |
+
+## Proposal 3: `use` API
+
+```ts
+export async function setup({ defineConfig, env }) {
+	const pluginJson = await env.$import(
+		"https://cdn.jsdelivr.net/gh/samuelstroschein/inlang-plugin-json@1/dist/index.js",
+	)
+	const sdk = await env.$import("https://cdn.jsdelivr.net/gh/inlang/inlang@1/dist/sdk-js/index.js")
+
+	return defineConfig({
+		referenceLanguage: "en",
+	})
+		.use(
+			pluginJson({
+				pathPattern: "./languages/{language}.json",
+			}),
+		)
+		.use(
+			sdk({
+				languageNegotiation: {
+					strategies: [{ type: "localStorage" }, { type: "navigator" }],
+				},
+			}),
+		)
+}
+```
+
+|                       Pros                       |                      Cons                      |
+| :----------------------------------------------: | :--------------------------------------------: |
+| Apps can read, extend and modify previous config |    Host applications must pass defineConfig    |
+|            No deep nesting of plugins            |                breaking change                 |
+|                                                  | `use` seems unnecessary compared to proposal 4 |
+
+Have a setup function that includes a defineConfig function. The return of `defineConfig` can be extended by calling `use` on it.
+
+A major downside is two-fold increase in complexity:
+
+1. Host applications must pass `defineConfig` to plugins, complicating the "consumption" of the inlang.config.js file.
+2. Consumers face two different APIs to define the config: `defineConfig` for core stuff and `use` for plugin stuff. That's a bit implicit.
+3. Consumers need to understand the concept of `setup` and `defineConfig`. In the current design `defineConfig` is the only setup function and concept that needs to be understood.
+
+## Proposal 4: `plugins` property
+
+```ts
+export async function defineConfig(env) {
+	const pluginJson = await env.$import(
+		"https://cdn.jsdelivr.net/gh/samuelstroschein/inlang-plugin-json@1/dist/index.js",
+	)
+	const sdk = await env.$import("https://cdn.jsdelivr.net/gh/inlang/inlang@1/dist/sdk-js/index.js")
+
+	const standardLintRules = await env.$import(
+		"https://cdn.jsdelivr.net/gh/inlang/inlang@1/dist/sdk-js/index.js",
+	)
+
+	return {
+		referenceLanguage: "en",
+		plugins: [
```

Comment:
the only downside I see is that a `plugin` can be anything. Maybe it is confusing to have a core functionality like reading and writing resources named "plugin"?

Reply:
Hm. I interpret this API as a plugin has a set of functions it can call. One of them is `defineConfig` which defines the config. Because a plugin defines the `readResources` and `writeResources` properties of the config, the user is not required to define those properties itself. 

Do you have an alternative proposal?

Reply:
the alternative is coming up with a better name for core functionality e.g.
```diff
return {
	referenceLanguage: "en",
-	plugins: [jsonPlugin],
+	ressources: [jsonRessources],
}
```
But `jsonRessoureces` would still be able to modify other properties on the config and not just `Ressource` related ones. This might also confuse people.
So not sure if it is worth coming up with another system.

Reply:
the alternative is coming up with a better name for core functionality e.g.
```diff
return {
	referenceLanguage: "en",
-	plugins: [jsonPlugin],
+	ressources: [jsonRessources],
}
```
But `jsonRessoureces` would still be able to modify other properties on the config and not just `Ressource` related ones. This might also confuse people.
So not sure if it is worth coming up with another system.

Reply:
It seems confusing/is another concept we need to explain. Thus, I refrain from further investigating `ressources`.